### PR TITLE
Routing rules overview doc

### PR DESCRIPTION
**Pull request by frankbu**
This change is [Reviewable](https://reviewable.io/reviews/istio/istio/28)

### CODE REVIEW

File: doc/rule-dsl.md [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+The `source` field is used to qualify a rule to only apply to requests from a specific caller.
+For example,
+the following `match` clause qualifies the rule to only apply to calls from the "reviews" microservice.
+
+```yaml
+destination: ratings.default.svc.cluster.local
+match:
+  source: reviews.default.svc.cluster.local
+```
+
```
Comment (rshriram) [inline doc/rule-dsl.md]: Also worth mentioning FQDN

File: doc/rule-dsl.md [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+to an incoming request if it includes a "Cookie" header that contains the substring "user=jason".
+
+```yaml
+destination: reviews.default.svc.cluster.local
+match:
+  http:
+    Cookie:
+      regex: "^(.*?;)?(user=jason)(;.*)?$"
+```
+
```
Comment (rshriram) [inline doc/rule-dsl.md]: You might also want to include the Uri field as an example (the most common one), with a prefix match.. like

```yaml
match:
   http:
     Uri:
       prefix: /svc/reviews/
```

File: doc/rule-dsl.md [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+* `weight`
+* `destination`
+
+Of these fields, `tags` is the only required one, the others are optional.
+
+#### Property: route.tags <a id="route-tags"></a>
+
+The `tags` field is a list of instance tags to identify the target instances (e.g., version) to route requests to.
+If there are multiple registered instances with the specified tag(s), they will be routed to in a round-robin fashion.
+
```
Comment (rshriram) [inline doc/rule-dsl.md]: requests will be load balanced among these instances based on the load balancing policy set in the Destination Policy

File: doc/rule-dsl.md [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+route:
+- tags:
+    version: v1
+http_req_timeout:
+  simple_timeout:
+    timeout_seconds: 10
+```
+
+Alternatively, you can specify a timeout override header that downstream clients can use to
+dynamically override the timeout on a per request basis:
```
Comment (rshriram) [inline doc/rule-dsl.md]: Please remove these two lines

File: doc/rule-dsl.md [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+- tags:
+    version: v1
+http_req_retries:
+  simple_retry:
+    attempts: 3
+```
+
+Just like timeouts, a retry override header can be specified to allow downstream clients to
+dynamically override the value on a per request basis:
+
```
Comment (rshriram) [inline doc/rule-dsl.md]: Please remove this as well, or alternatively, we can keep this text and link out to Envoy docs, that specify what the override header name is and the format of the value
https://lyft.github.io/envoy/docs/configuration/http_filters/router_filter.html#x-envoy-upstream-rq-timeout-ms

Either way, we cannot specify our own override header name with Envoy (it was included for the sake of other proxy implementations).

File: doc/rule-dsl.md [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+```yaml
+destination: "ratings.default.svc.cluster.local"
+route:
+- tags:
+    version: v1
+http_req_timeout:
+  simple_timeout:
+    override_header_name: My-Request-Timeout
+```
```
Comment (rshriram) [inline doc/rule-dsl.md]: This needs to be removed. But please see my comment below on the Envoy override headers (we can just link to envoy docs)

File: doc/rule-dsl.md [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+
+The optional `percent` field, a value between 0 and 100, is used to only abort a certain percentage of requests.
+All request are aborted by default.
+
+The `http_status` field is used to indicate a value to return from an HTTP request, instead of calling the backend.
+Its value is an integer HTTP 2xx, 3xx, 4xx, or 5xx status code.
+
+Similarly, to abort HTTP/2 or gRPC, the value to return 
+can be specified using the `http2_error` or `grpc_status`, respectively.
+
```
Comment (rshriram) [inline doc/rule-dsl.md]: Please add a note stating this feature is currently not implemented

File: doc/rule-dsl.md [hunk 8]
```diff
@@ -0,0 +1,510 @@
...
+  delay:
+    fixed_delay_seconds: 5
+  abort:
+    percent: 10
+    http_status: 400
+```
+
+### Route Rule Evaluation <a id="route-rule-evaluation"></a>
+
+Whenever the routing story for a particular microservice is purely weight base,
```
Comment (rshriram) [inline doc/rule-dsl.md]: typo: weight based
did you intend to say routing story ?

File: doc/rule-dsl.md [hunk 9]
```diff
@@ -0,0 +1,510 @@
...
+```
+
+#### Property: match.http <a id="match-http"></a>
+
+The `http` field is a set of one or more property-value pairs where each property is an HTTP header name
+and the corresponding value is one of the following:
+
+1. `exact: value`, where the header value must match `value` exactly
+2. `prefix: value`, where `value` must be a prefix of the header value
+3. `regex: value`, where `value` is a regular expression that matches the header value
```
Comment (kyessenov) [inline doc/rule-dsl.md]: nit: put value in quotes as it's likely to contain special characters.

File: doc/rule-dsl.md [hunk 10]
```diff
@@ -0,0 +1,510 @@
...
+Similar to route rules, destination policies are associated with a particular `destination` however
+if they also include `tags` their activation depends on route rule evaluation results.
+
+The first step in the rule evaluation process evaluates the route rules for a `destination`,
+if any are defined, to determine the tags (i.e., specific version) of the destination service
+that the current request will be routed to. Next, the set of destination policies, if any, are evaluated
+to determine if they apply.
+
+One subtlety of the algorithm to keep in mind is that policies that are defined for specific
+tagged destinations will only be applied if the corresponding tagged instances are explicity
```
Comment (kyessenov) [inline doc/rule-dsl.md]: Thanks for explaining this. It's indeed subtle.

File: doc/rule-dsl.md [hunk 11]
```diff
@@ -0,0 +1,510 @@
...
+for the service. For example, you can add a simple route rule for "reviews:v1".
+
+```yaml
+destination: reviews.default.svc.cluster.local
+route:
+- tags:
+    version: v1
+```
+
+Setting a default rule like this for every microservice is generally considered a best practice in Istio.
```
Comment (kyessenov) [inline doc/rule-dsl.md]: @rshriram It's probably worth mentioning the implicit default routing (all versions to all versions). That is by default, you can send traffic from any service to any other service.

**Comment by frankbu** [comment]
I will hold off on merging this until the bookinfo demo is updated to conform. Unfortunately bookinfo isn't working ATM with latest images - some kind of problem with ingress controller that needs to be debugged and fixed first

File: doc/rule-dsl.md (hunk 2 above)
Comment (frankbu) [inline doc/rule-dsl.md]: How does this work? Who sets this "Uri" header. Actually, this reminds me that we used to have a "uri" or "path" field in a previous version of the rule model. What happened to it, and how is this "uri" header related?

Comment (rshriram) [inline doc/rule-dsl.md]: Uri is set by Envoy

**Pull request merged**